examples/usb_i2c_spdm.rs

```diff
@@ -0,0 +1,174 @@
+//!
+//! For reception, the device (this app) first sends a packet header of the following format
+//!
+//! device_tx[0] = 0xBB // Preamble
+//! device_tx[1] = 0xFF // Unused
+//! device_tx[2] = YY   // MSB of 16bit data length
+//! device_tx[3] = ZZ   // LSB of 16bit data length
+//!
+//! Based on the data length, the host can read the next ((YY << 8) | ZZ )
+//! as data message.
+//!
+#![no_main]
+#![no_std]
+use libtock::console::Console;
+use libtock::i2c_master_slave::I2CMasterSlave;
+use libtock::leds::Leds;
+use libtock::runtime::{set_main, stack_size};
+
+set_main! {main}
+stack_size! {0x900}
+
+/// The address to which we listen for in slave/target mode.
+pub const MY_ID: u8 = 0x22;
+/// Contains packet metadata
+pub const HEADER_LEN: usize = 0x4;
+/// Max data message length
+pub const MAX_DLEN: usize = 128;
+/// Total amount of bytes we can receive in a single UART RX
+/// !! Note: When the tock kernel is built for the target, the UART_RX
+///          buffer size must be set to >= RX_BUF_LEN. Otherwise
+///          the UART driver will error out as expected due to
+///          the kernel RX buffer being smaller. Similarly,
+///          the I2CMasterSlave buffer sizes must also be
+///          >= MAX_DLEN.
+pub const RX_BUF_LEN: usize = HEADER_LEN + MAX_DLEN;
+/// Debug LED config, change these based on the board config
+pub const PANIC_LED: u32 = 1;
+/// Triggered when waiting for RX
+pub const RX_LED: u32 = 2;
+/// Triggered when TX in progress
+pub const TX_LED: u32 = 3;
+
+/// # Summary
+///
+/// A helper function to append the packet metadata do the outgoing buffer
+/// pointed to by @buf.
+///
+/// # Parameter
+///
+/// * `buf`: the buffer in which to append the metadata to.
+/// * `n_bytes`: length of data message body of the next message this is
+///              stored in the packet header, so that the host can
+///              determine the data message read length.
+///
+/// # Returns
+///
+/// Ok(n) returns the total size of bytes to send from this buffer (buf[0..n]).
+///
+/// # Panics
+///
+/// If the buffer capacity cannot fit the packet metadata
+///
+fn prepare_tx_header(buf: &mut [u8], n_bytes: usize) -> Result<usize, ()> {
+    let total_bytes_to_send = HEADER_LEN;
+    if total_bytes_to_send > buf.len() {
+        Leds::on(PANIC_LED).unwrap();
+        panic!("usb-i2c-header: buffer has no space");
+    }
+    // Setup Header
+    buf[0] = 0xBB;
+    buf[1] = 0xFF;
+    // Upper 8-bits
+    buf[2..=3].copy_from_slice(&u16::to_be_bytes(n_bytes as u16));
+
+    Ok(total_bytes_to_send)
+}
+
+fn main() {
+    if let Ok(leds_count) = Leds::count() {
+        // Make sure we have enough debugs
+        assert!(leds_count >= 3);
+    } else {
+        // Since we can't directly print error messages
+        // as the host is monitoring the UART bus, it's a good idea
+        // to use LEDs to indicate the current state of the device.
+        panic!("Not enough debug leds");
+    }
```

Comment:
This isn't ideal, there are a bunch of boards with not enough LEDs. Instead the LED blinking could be behind checks, for example

```diff
diff --git a/examples/usb_i2c_spdm.rs b/examples/usb_i2c_spdm.rs
index 8465cc5..0371ab1 100644
--- a/examples/usb_i2c_spdm.rs
+++ b/examples/usb_i2c_spdm.rs
@@ -55,11 +55,11 @@ pub const MAX_DLEN: usize = 128;
 ///          >= MAX_DLEN.
 pub const RX_BUF_LEN: usize = HEADER_LEN + MAX_DLEN;
 /// Debug LED config, change these based on the board config
-pub const PANIC_LED: u32 = 1;
+pub const PANIC_LED: u32 = 0;
 /// Triggered when waiting for RX
-pub const RX_LED: u32 = 2;
+pub const RX_LED: u32 = 1;
 /// Triggered when TX in progress
-pub const TX_LED: u32 = 3;
+pub const TX_LED: u32 = 2;
 
 /// # Summary
 ///
@@ -97,15 +97,8 @@ fn prepare_tx_header(buf: &mut [u8], n_bytes: usize) -> Result<usize, ()> {
 }
 
 fn main() {
-    if let Ok(leds_count) = Leds::count() {
-        // Make sure we have enough debugs
-        assert!(leds_count >= 3);
-    } else {
-        // Since we can't directly print error messages
-        // as the host is monitoring the UART bus, it's a good idea
-        // to use LEDs to indicate the current state of the device.
-        panic!("Not enough debug leds");
-    }
+    let leds_count = Leds::count().unwrap_or(0);
+
     // RX Buffer layout
     // [0] = Preamble
     // [1] = TargetID
@@ -116,12 +109,18 @@ fn main() {
     let mut target_id: u16;
 
     loop {
-        Leds::on(RX_LED).unwrap();
+        if leds_count > RX_LED {
+            Leds::on(RX_LED).unwrap();
+        }
         let (_, err) = Console::read(&mut rx_buf);
-        Leds::off(RX_LED).unwrap();
+        if leds_count > RX_LED {
+            Leds::off(RX_LED).unwrap();
+        }
 
         if err.is_err() {
-            Leds::on(PANIC_LED).unwrap();
+            if leds_count > PANIC_LED {
+                Leds::on(PANIC_LED).unwrap();
+            }
             panic!("Failed to read from host {:?}", err);
```

Reply:
Fixed this with led_on/off closures